src/Microsoft.Framework.Caching.SqlServer/SqlServerCache.cs

```diff
@@ -0,0 +1,478 @@
+            _sqlQueries = new SqlQueries(_options.TableName);
+            _logger = loggerFactory.CreateLogger<SqlServerCache>();
+        }
+
+        public void Connect()
+        {
+            // Try connecting to the database to test if its available and also
+            // get and cache the ordinal positions of the columns for better performance.
+            using (var connection = new SqlConnection(_options.ConnectionString))
+            {
+                var command = new SqlCommand(_sqlQueries.GetTableSchema, connection);
+                connection.Open();
+
+                var reader = command.ExecuteReader(CommandBehavior.SchemaOnly);
+
+                _cacheItemIdIndex = reader.GetOrdinal(CacheItemId);
+                _cacheItemValueIndex = reader.GetOrdinal(CacheItemValue);
+                _expiresAtTimeUTCIndex = reader.GetOrdinal(ExpiresAtTimeUTC);
+                _slidingExpirationInTicksIndex = reader.GetOrdinal(SlidingExpirationInTicks);
+                _absoluteExpirationIndex = reader.GetOrdinal(AbsoluteExpiration);
+            }
+        }
+
+        public async Task ConnectAsync()
+        {
+            // Try connecting to the database to test if its available and also
+            // get and cache the ordinal positions of the columns for better performance.
+            using (var connection = new SqlConnection(_options.ConnectionString))
+            {
+                var command = new SqlCommand(_sqlQueries.GetTableSchema, connection);
+                await connection.OpenAsync();
+
+                var reader = await command.ExecuteReaderAsync(CommandBehavior.SchemaOnly);
+
+                _cacheItemIdIndex = reader.GetOrdinal(CacheItemId);
+                _cacheItemValueIndex = reader.GetOrdinal(CacheItemValue);
+                _expiresAtTimeUTCIndex = reader.GetOrdinal(ExpiresAtTimeUTC);
+                _slidingExpirationInTicksIndex = reader.GetOrdinal(SlidingExpirationInTicks);
+                _absoluteExpirationIndex = reader.GetOrdinal(AbsoluteExpiration);
+            }
+        }
+
+        public byte[] Get([NotNull] string key)
+        {
+            byte[] cacheItemvalue = null;
+            var utcNow = _systemClock.UtcNow;
+
+            using (var connection = new SqlConnection(_options.ConnectionString))
+            {
+                var command = new SqlCommand(_sqlQueries.GetCacheItem, connection);
+                command.Parameters.AddWithValue(CacheItemId, key);
+                command.Parameters.AddWithValue("UtcNow", utcNow);
+
+                connection.Open();
+
+                var reader = command.ExecuteReader(CommandBehavior.SingleRow);
+
+                DateTimeOffset oldExpirationTime;
+                TimeSpan? slidingExpiration = null;
+                DateTimeOffset? absoluteExpiration = null;
+                if (reader.Read())
+                {
+                    oldExpirationTime = reader.GetFieldValue<DateTimeOffset>(_expiresAtTimeUTCIndex);
```

Comment:
Since we have the optimization in place in which we will double the lifetime of the entry, we could also check here if we are getting an already expired item (based on the period that was actually specified, not the doubled one) and decide to return null, plus delete the item.